File: netlogo-gui/src/main/prim/_fd.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.core.Let;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.MutableDouble;

// note that this and _bk are pretty much carbon copies of each other

public final strictfp class _fd
    extends Command
    implements org.nlogo.nvm.CustomAssembled {

  final Let let;

  public _fd(org.nlogo.core.Token token) {
    super();
    token_$eq(token);
    let = new Let("~" + token.text() + "_" + token.start());
  }

  @Override
  public void perform(Context context) throws LogoException {
    perform_1(context, argEvalDoubleValue(context, 0));
  }

  public void perform_1(Context context, double d) {
    context.activation.binding.let(let, new MutableDouble(d));
    context.ip = next;
  }

  public void assemble(org.nlogo.nvm.AssemblerAssistant a) {
    a.add(this);
    a.add(new _fdinternal(this));
  }
}

File: netlogo-headless/src/main/prim/_fd.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.core.Let
import org.nlogo.nvm.{ AssemblerAssistant, Command, Context, CustomAssembled, MutableDouble }

// note that this and _bk are pretty much carbon copies of each other

class _fd extends Command with CustomAssembled {

  // MethodRipper won't let us call a public method from perform_1() - ST 7/20/12
  private[this] val _let = Let()
  def let = _let

  override def perform(context: Context) {
    perform_1(context, argEvalDoubleValue(context, 0))
  }

  def perform_1(context: Context, d: Double) {
    context.let(_let, new MutableDouble(d))
    context.ip = next
  }

  def assemble(a: AssemblerAssistant) {
    a.add(this)
    a.add(new _fdinternal(this))
  }

}
